src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -872,64 +922,112 @@ CreateDistributedExecution(RowModifyLevel modLevel, List *taskList, bool hasRetu
- * accesses to enforce restrictions that arise due to foreign keys to reference
- * tables.
+ * InitialiseCoordinatedTransactionForTaskList decides whether to initiate a coordinated
+ * transaction and whether to use 2PC for the given task list. These are global settings
+ * that affect the remainder of the transaction.
+ *
+ * Additionally, it returns whether the current execution should error out on any failure
+ * and whether to use transaction blocks, which may depend on previous statements in the
+ * transaction.
  */
-void
-StartDistributedExecution(DistributedExecution *execution)
+static void
+InitialiseCoordinatedTransactionForTaskList(RowModifyLevel modLevel, List *taskList,
+											bool *errorOnAnyFailure,
+											TransactionBlocksUsage *useTransactionBlocks)
 {
-	List *taskList = execution->tasksToExecute;
+	*errorOnAnyFailure = false;
+	*useTransactionBlocks = TRANSACTION_BLOCKS_ALLOWED;
+
+	if (taskList == NIL)
+	{
+		/* nothing to do */
+		return;
+	}
 
-	if (MultiShardCommitProtocol != COMMIT_PROTOCOL_BARE)
+	if (MultiShardCommitProtocol == COMMIT_PROTOCOL_BARE)
 	{
 		/*
-		 * In case localExecutionHappened, we simply force the executor to use 2PC.
-		 * The primary motivation is that at this point we're definitely expanding
-		 * the nodes participated in the transaction. And, by re-generating the
-		 * remote task lists during local query execution, we might prevent the adaptive
-		 * executor to kick-in 2PC (or even start coordinated transaction, that's why
-		 * we prefer adding this check here instead of
-		 * Activate2PCIfModifyingTransactionExpandsToNewNode()).
+		 * We prefer to error on any failures for CREATE INDEX
+		 * CONCURRENTLY or VACUUM//VACUUM ANALYZE (e.g., COMMIT_PROTOCOL_BARE).
 		 */
-		if (DistributedExecutionRequiresRollback(execution) || LocalExecutionHappened)
-		{
-			BeginOrContinueCoordinatedTransaction();
+		*errorOnAnyFailure = true;
+		*useTransactionBlocks = TRANSACTION_BLOCKS_DISALLOWED;
+		return;
+	}
 
-			if (TaskListRequires2PC(taskList) || LocalExecutionHappened)
-			{
-				/*
-				 * Although using two phase commit protocol is an independent decision than
-				 * failing on any error, we prefer to couple them. Our motivation is that
-				 * the failures are rare, and we prefer to avoid marking placements invalid
-				 * in case of failures.
-				 */
-				CoordinatedTransactionUse2PC();
+	/*
+	 * In case localExecutionHappened, we force the executor to use 2PC.
+	 * The primary motivation is that at this point we're definitely expanding
+	 * the nodes participated in the transaction. And, by re-generating the
+	 * remote task lists during local query execution, we might prevent the adaptive
+	 * executor to kick-in 2PC (or even start coordinated transaction, that's why
+	 * we prefer adding this check here instead of
+	 * Activate2PCIfModifyingTransactionExpandsToNewNode()).
+	 */
+	if (LocalExecutionHappened)
+	{
+		BeginOrContinueCoordinatedTransaction();
```

Comment:
Maybe I am missing something but to me it seems kind of counter intuitive that we start a transaction when `useTransactionBlocks` is not required then later we set it to `REQUIRED`. Shouldnt the reason we are starting the transaction be the other way around?

Reply:
Prior to this function we are undecided on the value of useTransactionBlocks. Is the main issue that we set useTransactionBlocks to ALLOWED at the start of this function? I could move it from there.

Reply:
No I am not talking about setting it at the beginning of function.

If useTransactionBlocks is required then we should start a coordinated transaction. So intuitively I would expect that it is set to required before this function is called, not after. Does that make sense?

Reply:
Is  that we should not call `BeginOrContinueCoordinatedTransaction()`/`CoordinatedTransactionUse2PC()` in `DecideTransactionProperties()`,  but on `StartDistributedExecution()`? But then we should also get 1PC/2PC from this function as well, which makes sense btw.